cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -457,12 +457,38 @@ def send_data(tls):
         'kubernetes.default.svc',
         'kubernetes.default.svc.{0}'.format(domain)
     ]
+
+    # maybe they have extra names they want as SANs
+    extra_sans = hookenv.config('extra_sans')
+    if extra_sans and not extra_sans == "":
+      sans.extend(extra_sans.split())
+
     # Create a path safe name by removing path characters from the unit name.
     certificate_name = hookenv.local_unit().replace('/', '_')
     # Request a server cert with this information.
     tls.request_server_cert(common_name, sans, certificate_name)
 
 
+@when('config.changed', 'certificates.available')
+def update_certificate(tls):
+    # I using the config.changed flag instead of something more
+    # specific to try and catch ip changes. Being a little
+    # spammy here is ok because the cert layer checks for
+    # changes to the cert before issuing a new one
+    send_data(tls)
```

Comment:
I think we can omit this handler. `send_data` is already invoked on every update-status hook (every ~5 min):

```
$ juju ssh kubernetes-master/0 sudo cat /var/log/juju/unit-kubernetes-master-0.log | grep send_data | tail -n 5
2017-10-19 21:12:58 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
2017-10-19 21:17:19 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
2017-10-19 21:22:42 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
2017-10-19 21:27:24 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
2017-10-19 21:32:10 INFO juju-log Invoking reactive handler: reactive/kubernetes_master.py:437:send_data
```

`config.changed` wouldn't cover the ip change case, either. It's only set if the application config changes (e.g. `juju config kubernetes-master ...`). I don't think there is a state that captures an IP change, but considering that `send_data` runs frequently already, we're probably good.

Reply:
My understanding from @johnsca was that an ip change would trigger config.changed. If that isn't the case then this is just hitting for no good reason when another config option is changed and I agree it should be removed. I worry that a possible 5 minute lag before you can hit your master isn't good though. At a minimum, I would like to change that to config.changed.extra_sans so we kick it off when you change your SAN config.

Reply:
I'm +1 on changing this to config.changed.extra_sans. I don't think we can rely on juju to tell us when the IP has changed, but you could always add the IP to extra_sans manually if you didn't want to wait the 5 minutes for the next update-status hook.